core/src/Telemetry/TelemetryManager.cs

```diff
@@ -0,0 +1,243 @@
+        }
+
+        internal void StopEvent(string requestId, EventBase eventToStop)
+        {
+            if (!HasReceiver() || string.IsNullOrWhiteSpace(requestId))
+            {
+                return;
+            }
+
+                var eventKey = new Tuple<string, string>(requestId, eventToStop[EventBase.EventNameKey]);
+
+                // Locate the same name event in the EventsInProgress map
+                EventBase eventStarted = null;
+                if (EventsInProgress.ContainsKey(eventKey))
+                {
+                    eventStarted = EventsInProgress[eventKey];
+                }
+
+                // If we did not get anything back from the dictionary, most likely its a bug that StopEvent
+                // was called without a corresponding StartEvent
+                if (null == eventStarted)
+                {
+                    // Stop Event called without a corresponding start_event.
+                    return;
+                }
+
+                // Set execution time properties on the event
+                eventToStop.Stop();
+
+                if (!CompletedEvents.ContainsKey(requestId))
+                {
+                    // if this is the first event associated to this
+                    // RequestId we need to initialize a new List to hold
+                    // all of sibling events
+                    var events = new List<EventBase>
+                    {
+                        eventToStop
+                    };
+                    CompletedEvents[requestId] = events;
+                }
+                else
+                {
+                    // if this event shares a RequestId with other events
+                    // just add it to the List
+                    CompletedEvents[requestId].Add(eventToStop);
+                }
+
+                // Mark this event as no longer in progress
+                EventsInProgress.TryRemove(eventKey, out var dummy);
+        }
+
+        internal void Flush(string requestId, string clientId)
+        {
+            if (!HasReceiver())
+            {
+                return;
+            }
+
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // No completed Events returned for RequestId
+                return;
+            }
+
+            CompletedEvents[requestId].AddRange(CollateOrphanedEvents(requestId));
+            CompletedEvents.TryRemove(requestId, out List<EventBase> eventsToFlush);
+
+            bool onlySendFailureTelemetry;
+            lock (_lockObj)
+            {
+                onlySendFailureTelemetry = _telemetryReceiver?.OnlySendFailureTelemetry ?? false;
+            }
+
+            if (onlySendFailureTelemetry)
+            {
+                // iterate over Events, if the ApiEvent was successful, don't dispatch
+                bool shouldRemoveEvents = false;
+
+                foreach (var anEvent in eventsToFlush)
```

Comment:
Nit: possibly easier to read: eventsToFlush.Any(ev => ev is ApiEvent a && a.WasSuccessful)

Reply:
I played with various cleanups of this and never got one I liked. This is exactly what I want.  Will fix.